src/python/pants/backend/codegen/protobuf/protobuf_dependency_inference.py

```diff
@@ -12,22 +16,30 @@
 from pants.engine.fs import Digest, DigestContents
 from pants.engine.rules import Get, MultiGet, collect_rules, rule
 from pants.engine.target import (
+    Dependencies,
+    DependenciesRequest,
+    ExplicitlyProvidedDependencies,
     HydratedSources,
     HydrateSourcesRequest,
     InferDependenciesRequest,
     InferredDependencies,
     SourcesPathsRequest,
     Targets,
+    WrappedTarget,
 )
 from pants.engine.unions import UnionRule
 from pants.util.frozendict import FrozenDict
 from pants.util.logging import LogLevel
-from pants.util.ordered_set import FrozenOrderedSet
+from pants.util.ordered_set import FrozenOrderedSet, OrderedSet
 
 
-class ProtobufMapping(FrozenDict[str, Address]):
+@dataclass(frozen=True)
+class ProtobufMapping:
     """A mapping of stripped .proto file names to their owning file address."""
 
+    mapping: FrozenDict[str, Address]
```

Comment:
It is. Marking the dataclass as frozen means that dataclass with autogenerate __hash__ and __eq__ for us: it's the spiritual equivalent of `#[derive(Hash, PartialEq)]`. 

For the __hash__ to actually work at runtime, every field needs to be hashable. We need to use FrozenDict instead of `dict` - `dict` isn't hashable because it's mutable. Same with `tuple` instead of `list`.